### ISSUE TITLE
Summarizing timestamps by month handles time zones incorrectly in Snowflake

### ISSUE THREAD

**Issue by gouline** [metabase/metabase#37065]
### Describe the bug

When summarizing timestamps by month in Snowflake, the generated SQL incorrectly buckets the same month multiple times for each time zone present in the data, instead of aggregating all rows in that month relative to the current time zone.

This is evident in DST changeover months, for example when Sydney went from GMT+10 to GMT+11 on October 1. Summarize question creates two buckets for October: before (starting at midnight GMT+10) and after (starting at midnight GMT+11). 

**This is a regression in x.48.0 as x.47.9 works correctly.**

Out of Snowflake and Postgres that I tested on, this only happens on Snowflake, however it may affect other databases.

### To Reproduce

Create fake data in Snowflake:

```sql
alter session set timezone = 'Australia/Sydney';
create table if not exists tz_test (ts timestamp_tz);
insert into tz_test(ts) values
    ('2023-09-30 23:59:59 +1000'), // September
    ('2023-10-01 00:00:00 +1000'), // October before DST starts
    ('2023-10-01 05:00:00 +1100'), // October after DST starts
    ('2023-11-01 05:00:00 +1100'); // November
```

Now create a Metabase question for that table to summarize "Count" by "Ts: Month".

### Result

| Ts: Month       | Count |
|-----------------|-------|
| September, 2023 | 1     |
| October, 2023   | 2     |
| November, 2023  | 1     |

### SQL

Note `AS timestamp` appended to timestamps.

```sql
SELECT
  DATE_TRUNC("month", CAST("PUBLIC"."TZ_TEST"."TS" AS timestamp)) AS "TS",
  COUNT(*) AS "count"
FROM
  "TZ_TEST"."PUBLIC"."TZ_TEST"
GROUP BY
  DATE_TRUNC("month", CAST("PUBLIC"."TZ_TEST"."TS" AS timestamp))
ORDER BY
  DATE_TRUNC("month", CAST("PUBLIC"."TZ_TEST"."TS" AS timestamp)) ASC
```

### Executing SQL directly in Snowflake

| TS                      | count |
|-------------------------|-------|
| 2023-09-01 00:00:00.000 | 1     |
| 2023-10-01 00:00:00.000 | 2     |
| 2023-11-01 00:00:00.000 | 1     |

### Result

| Ts: Month       | Count |
|-----------------|-------|
| September, 2023 | 1     |
| September, 2023 | 1     |
| October, 2023   | 1     |
| November, 2023  | 1     |

### SQL

Note `AS timestamp` is missing.

```sql
SELECT
  DATE_TRUNC("month", "PUBLIC"."TZ_TEST"."TS") AS "TS",
  COUNT(*) AS "count"
FROM
  "TZ_TEST"."PUBLIC"."TZ_TEST"
GROUP BY
  DATE_TRUNC("month", "PUBLIC"."TZ_TEST"."TS")
ORDER BY
  DATE_TRUNC("month", "PUBLIC"."TZ_TEST"."TS") ASC
```

### Executing SQL directly in Snowflake

| TS                            | count |
|-------------------------------|-------|
| 2023-09-01 00:00:00.000 +1000 | 1     |
| 2023-10-01 00:00:00.000 +1100 | 1     |
| 2023-10-01 00:00:00.000 +1000 | 1     |
| 2023-11-01 00:00:00.000 +1100 | 1     |

### Expected behavior

0.47.9 behavior is expected

### Information about your Metabase installation

```JSON
- Metabase version: 0.48.0 vs 0.47.9
- Target database: Snowflake
```

### Severity

Breaks all month-on-month questions with time series data

### Additional context

![Screenshot 2023-12-22 at 18 14 11](https://github.com/metabase/metabase/assets/1960272/67bffd88-b6fb-4013-92ed-b6775d88c828)

**Comment by zbodi74** [comment]
There are multiple customers affected. Here's a reproduction to illustrate how the issue leads to misleading results.
1. Create a native query: 
```
--- returns one record per day for year 2024
SELECT 
 DATEADD('DAY', x, '2024-01-01T00:00:00Z'::TIMESTAMP_TZ) AS created_at,
 x+1 AS day_of_year
FROM (
  SELECT seq4() x FROM TABLE(GENERATOR(ROWCOUNT => 366))
)
```
2. Explore results, summarize: Count of rows, grouped by CREATED_AT:Month
This should show the number of days in each month in the year 2024.

3. Set report timezone to 'database default', and results are correct:
<img width="210" alt="image" src="https://github.com/metabase/metabase/assets/11080638/0bf0d890-ab24-45c1-8f2d-ec64a92895de">

4. Set report timezone to 'America/Los_Angeles', and get incorrect and misleading results.
Note that the count values are correct, but the month grouping is off by one.
<img width="203" alt="image" src="https://github.com/metabase/metabase/assets/11080638/aedb3dcb-2dc6-416e-9086-770ea0150ebd">

`/api/dataset` response:
```
{
    "data": {
        "rows": [
            ["2023-12-31T16:00:00-08:00", 31],
            ["2024-01-31T16:00:00-08:00", 29],
            ["2024-02-29T16:00:00-08:00",31],
[...]
            ["2024-11-30T16:00:00-08:00", 31]
        ],
```

**Comment by metamben** [comment]
#35574 might be the change that introduced the regression. In 48.4 we released #36413, also changing timezone handling.

**Comment by metamben** [comment]
Tested with report timezone set to Australia/Sydney and the native query:

```sql
select t tz, t::timestamp ntz, date_trunc('MONTH', t) mtz, date_trunc('MONTH', t::timestamp) mntz from (
      select ('2023-09-30 23:59:59 +1000')::timestamptz as t // September
union select ('2023-10-01 00:00:00 +1000')::timestamptz as t // October before DST starts
union select ('2023-10-01 05:00:00 +1100')::timestamptz as t // October after DST starts
union select ('2023-11-01 05:00:00 +1100')::timestamptz as t // November
) tabla
```

This shows the problem that Snowflake truncates the value and then shifts it according to the time zone, producing a second September value different from the first one.
